### PR TITLE
Use the project's build path defined Java version as the default

### LINKED ISSUE #271: Can we offer the project-specific compiler JDK as an option in the Run Config?
We have three choices in the Run Config:
1. Workspace default
2. Execution env
3. Alternate JRE

Could we also offer a choice like "Project-specific" compiler?  Would this be a better default than workspace default?   Should the default for the run config be a combination of project-specific compiler / workspace default?  

From another angle:   is there a way for Liberty Tools Eclipse to make it so that, if you import a Java 8 project, once you configure a Java 8 JRE it automatically gets used for the Run Config?

### PR DESCRIPTION
**Pull request by mezarin**
Fixes #271

### CODE REVIEW

**Review by scottkurz — COMMENTED** [review]
## MAIN ISSUE

This works nicely with a single module with the plain "Start" command. 

It doesn't seem to work as expected with the custom "Start..." though.   There's more work to be done with the lifecycle and/or defaulting of the Run Config.   

But I think you can see what's going on if you just click on the "JRE" tab.  You'll see the default is set as if the workspace JRE ('jre') were selected as an "Alternate JRE".

image

It seems that once the Run/Debug config editor window is active, this default gets set and it ends up taking precedence over the path where we default to the value obtained from reading the .classpath file (through the API).    I don't think you necessarily need to do "Start ..."...just open the Run Configuration and do something like click in the start param text box field.

Maybe this is an easy to thing to fix.  Although it reminds me of how I struggled with this API myself when I was trying to enable modifying the project a Run/Debug config was associated with, when first defaulting to the active project.  So I appreciate if it turns out this isn't easy.  

## OTHER ISSUES

1. Consider the case where we're running this from a non-Java module... e.g. from the EAR or top-level project in a Maven multi-mod.    (Yes, sometimes this will be imported into Eclipse with Java nature and sometimes not.... still haven't totally figured out exactly when vs. not.)  

     Anyway... this reminds us of the need to find a Java project to hold the debugger config.  Could it be worth going through the same code path to find a Java project to find the default Java level??   

    My intuition is that, in the narrow sense of helping provide a better default for this particular usage, the value of making that change is kind of small.   However it may be valuable to continue building upon the precedent:  whenever we need to look at a Java setting or construct we go through the "find me a Java module" routine.    This might be a more orderly way to go forwards, so might be worth it.

2. Are we sure we want to change the key we use within the run config attribute map?  

    We were using:   `public static final String JRE_CONTAINER_KEY = "org.eclipse.jdt.launching.JRE_CONTAINER";`

    True, we're still in "Early Release", i.e. "beta", so we have more room to make incompatible changes.  I'm just wondering could this basically have been done using the same attribute key, but you changed because you think the name is a better fit?   Or is there more of a reason to leave the past behind and move to a new attribute... e.g. we want to go through the "recalculate the default" process again? 

    It's probably a good exercise to go through in any case, (to reason through whether we really need to break the existing config).   Note the configs are serialized in files like: `<workspace>/.metadata/.plugins/org.eclipse.debug.core/.launches/MyConfig.launch` or you can do "Export".

## AUTOMATED TESTS

It seems we need something here.  Though it would add a complicating layer to our tests to have to deal with a whole other JDK installed.   

If the work we need to do to address the JRE Tab default (above) suggests the need for a different default value that we could read and validate.. maybe there is some value pursuing that...and we wouldn't even necessarily have to actually start/run anything.   E.g. we could have a Java 8 compile setting in the POM and just validate that we get a corresponding JRE Tab setting on the Run Config.   But we don't necessarily have to actually run w/ Java 8 to validate.

## TRIVIAL ISSUES

1. getDefaulJavaFromBuildPath => getDefaultJavaFromBuildPath

**Comment by mezarin** [comment]
Thanks for taking a look.

**Main issue:** 

I don't see the issue you are seeing. If i open the Liberty tools configuration JRE tab for a project, the JRE set matches the project's .classpath JRE entry, and it also matches the Java build path as seen in the UI.
image
image

**Other issues:**

   1. We should have a discussion about that.
   2. Not sure i follow this comment. Both are exactly the same. The change just replaces our local representation/copy of the configuration attribute constant key with the one already defined in JDT.
  
**Automated tests and trivial issues:**

  I will look into adding tests for this and will fix that  typo.

**Comment by scottkurz** [comment]
Maybe we need to each run with some trace enabled to understand the different observations here...

For the "2." comment:
    >...Both are exactly the same. The change just replaces our local representation/copy of the configuration attribute constant key with the one already defined in JDT.

I missed that it's the same key String value so ignore that.

**Review by scottkurz — APPROVED** [review]
This looks good.   I did give it a pretty good review, but don't have much to note.

The multi-mod case where the aggregate is a non-Java is handled well.

The test looks like a good way to handle it.

I see the one test change you made to break out of the nested inner loop (by adding a new flag).

---

One thing to note:   a Maven project without an explicit default Java level will still pick one up from the "super POM"... I think it will use 1.5.    I'm not sure how to rationalize the equivalent in Gradle but I think the .classpath file will get written with a specific version of Java.

In other words, I don't think we'll ever default to the workspace default, which is fine...just noting.

Thanks for all the work on this.
